File: akka-docs/src/test/java/jdocs/stream/operators/flow/StatefulMap.java
/*
 * Copyright (C) 2022 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.stream.operators.flow;

import akka.actor.typed.ActorSystem;
import akka.japi.Pair;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Source;

import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;

public class StatefulMap {
    static final ActorSystem<?> system = null;

    public void indexed() {
        //#indexed
        Source.from(Arrays.asList("A", "B", "C", "D"))
            .statefulMap(() -> 0L,
                (index, element) -> Pair.create(index + 1, Pair.create(index, element)),
                Optional::ofNullable)
            .runForeach(System.out::println, system);
        //prints
        //Pair(0,A)
        //Pair(1,B)
        //Pair(2,C)
        //Pair(3,D)
        //#indexed
    }

    public void bufferUntilChanged() {
        // #bufferUntilChanged
        Source.from(Arrays.asList("A", "B", "B", "C", "C", "C", "D"))
            .statefulMap(() -> (List<String>) new LinkedList<String>(),
                (list, element) -> {
                    if (list.isEmpty()) {
                        list.add(element);
                        return Pair.create(list, Collections.<String>emptyList());
                    } else if (list.get(0).equals(element)) {
                        list.add(element);
                        return Pair.create(list, Collections.<String>emptyList());
                    } else {
                        final List<String> newBuffer = new LinkedList<>();
                        newBuffer.add(element);
                        return Pair.create(newBuffer, list);
                    }
                }, Optional::ofNullable)
            .runForeach(System.out::println, system);
        //prints
        //[A]
        //[B, B]
        //[C, C, C]
        //[D]
        //#bufferUntilChanged
    }

    public void distinctUntilChanged() {
        //#distinctUntilChanged
        Source.from(Arrays.asList("A", "B", "B", "C", "C", "C", "D"))
            .statefulMap(Optional::<String>empty,
                (lastElement, element) -> {
                    if (lastElement.isPresent() && lastElement.get().equals(element)) {
                        return Pair.create(lastElement, Optional.<String>empty());
                    } else {
                        return Pair.create(Optional.of(element), Optional.of(element));
                    }
                }, Optional::ofNullable)
            .via(Flow.flattenOptional())
            .runForeach(System.out::println, system);
        //prints
        //A
        //B
        //C
        //D
        //#distinctUntilChanged
    }

    public void usingResource() {
        // #usingResource
        Source.from(Arrays.asList("A", "B", "C", "D"))
            .statefulMap(() -> {
                //allocates some kind of new resource
                final AtomicInteger resource = new AtomicInteger(0);
                resource.set(100);
                return resource;
            }, (resource, element) -> {
                //mock resource usage
                final String output = "element " + element + " use resource: " + resource.getAndDecrement();
                return Pair.create(resource, output);
            }, r -> {
                //release resource
                r.set(0);
                return Optional.empty();
            }).runForeach(System.out::println, system);
        //prints
        //element A use resource: 100
        //element B use resource: 99
        //element C use resource: 98
        //element D use resource: 97
        //#usingResource
    }
}

File: akka-docs/src/test/scala/docs/stream/operators/flow/StatefulMap.scala
/*
 * Copyright (C) 2022 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.stream.operators.flow
import akka.actor.ActorSystem
import akka.stream.scaladsl.Sink
import akka.stream.scaladsl.Source

import java.util.concurrent.atomic.AtomicInteger

object StatefulMap {

  implicit val actorSystem: ActorSystem = ???

  def indexed(): Unit = {
    //#indexed
    Source(List("A", "B", "C", "D"))
      .statefulMap(() => 0L)((idx, elem) => (idx + 1, (idx, elem)), _ => None)
      .runWith(Sink.foreach(println))
    //prints
    //(0,A)
    //(1,B)
    //(2,C)
    //(3,D)
    //#indexed
  }

  def bufferUntilChanged(): Unit = {
    //#bufferUntilChanged
    Source("A" :: "B" :: "B" :: "C" :: "C" :: "C" :: "D" :: Nil)
      .statefulMap(() => List.empty[String])(
        (buffer, elem) =>
          buffer match {
            case Nil                       => (elem :: Nil, Nil)
            case head :: _ if head != elem => (elem :: Nil, buffer.reverse)
            case _                         => (elem :: buffer, Nil)
          },
        buffer => Some(buffer.reverse))
      .filter(_.nonEmpty)
      .runWith(Sink.foreach(println))
    //prints
    //List(A)
    //List(B, B)
    //List(C, C, C)
    //List(D)
    //#bufferUntilChanged
  }

  def distinctUntilChanged(): Unit = {
    //#distinctUntilChanged
    Source("A" :: "B" :: "B" :: "C" :: "C" :: "C" :: "D" :: Nil)
      .statefulMap(() => Option.empty[String])(
        (lastElement, elem) =>
          lastElement match {
            case Some(head) if head == elem => (Some(elem), None)
            case _                          => (Some(elem), Some(elem))
          },
        _ => None)
      .collect { case Some(elem) => elem }
      .runWith(Sink.foreach(println))
    //prints
    //A
    //B
    //C
    //D
    //#distinctUntilChanged
  }

  def usingResource(): Unit = {
    //#usingResource
    def allocateResource: AtomicInteger = {
      //allocates some kind of new resource
      val resource = new AtomicInteger(0)
      resource.set(100)
      resource
    }
    Source(List("A", "B", "C", "D"))
      .statefulMap(() => allocateResource)(
        (resource, elem) => {
          //use the resource
          (resource, s"element $elem use resource: " + resource.getAndDecrement())
        },
        resource => {
          //release resource
          resource.set(0)
          None
        })
      .runWith(Sink.foreach(println))
    //prints
    //element A use resource: 100
    //element B use resource: 99
    //element C use resource: 98
    //element D use resource: 97
    //#usingResource
  }

}
